lints/type_cosplay/src/lib.rs

```diff
@@ -61,31 +64,59 @@ struct TypeCosplay {
     deser_types: FxHashMap<AdtKind, Vec<(DefId, Span)>>,
 }
 
+// get type X
+// check if implements Discriminator
+// check corresponding function call type:
+// if !try_deserialize
+// emit lint with warning to use try_deserialize (because any types deriving Discriminator should since it guaranteed to check discrim)
+
 impl<'tcx> LateLintPass<'tcx> for TypeCosplay {
     fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>) {
         if_chain! {
-            if !expr.span.from_expansion();
-            if let ExprKind::Call(fnc_expr, args_exprs) = expr.kind;
-            if is_deserialize_function(cx, fnc_expr);
+            if let ExprKind::Call(fnc_expr, _args_exprs) = expr.kind;
+            // TODO: is the following if statement really needed?? don't think it's ever used.
```

Comment:
@smoelius this is something you might want to confirm. I commented it out for the time being. Basically, the if statement on lines 71-73 walks the argument to X::deser(arg), and check that there is a data reference on AccountInfo, ie, `AccountInfo.data` somewhere. I don't remember why I thought I needed this. But encountering it for the 2nd time, it seems unnecessary since what bytes we deserialize from shouldn't impact whether the code will have a type-cosplay issue or not...

Reply:
update: putting this check of the function arguments back in. Reason is, currently I get the type X, check if it implements Discriminator, then check the associated function call in X::call(), and then check if call() == try_deserialize(). This can lead to a bunch of false positives, since there could be many calls on X, not having to do with deserialization at all. I can't remember if this was my original intention, but having the check that args references `AccountInfo.data` basically gives more indication that this function call _is_ more likely to be a deserialization call.

That raises another issue though. I can't think of a specific example, but I feel like programs don't necessarily have to deserialize from the `AccountInfo.data` bytes...it could even be a reference or a borrow. So maybe we make this check less restrictive and just check that the argument to the call has type array of bytes? (but then again it could be wrapped in a RefCell, Rc, etc., idk if that matters)

Reply:
update 2: took it out again. reason is because it makes the recommended test pass for the wrong reason, and bypasses a ton of other necessary checks to actually check type-cosplay. In the recommended test, the argument to `try_deserialize` does not contain a reference to `AccountInfo.data`, it's just a byte array essentially.

So for now, this is taken out and it may yield some false positives. I modified the error message to better help guide the user in that case. A future improvement could be adding this check and improving it like I described above.